# PHASE I DATASCIENCE PROJECT

#                         AVIATION ACCIDENTS AND RISK FACTORS ANALYSIS 


* Student name: LILIAN KABURO

* Student pace: SELF PACED/HYBRID

* Instructor name: MARYANN MWIKALI



# 1. Introduction



Despite the number of aviation accidents recorded over the years,flying remains the safest mode of transport.In comparison to accidents and fatalities from let's say road transport,aviation accidents are not as frequent.That said,aviation industry is a lucrative venture but not without its risks and limitations.This projects aims to identify risk factors that contribute to accident occurrences and how they relate to specific aircraft models to assist company stakeholders make informed decisions on which aircraft model is suited best for new venture into the aviation industry.



# 1(a) Project Overview


# Objective

This project,using the data from NTSB Aviation Accidents dataset(1962-2023), aims to assess and identify the aircraft model with the lowest risk factors that affect and contribute to the frequency and severity of aviation accidents.
The information derived from an extensive analysis of the available data will help in making informed decisions in  purchasing the aircraft to venture into the aviation industry.

#  Business Understanding / Problem Definition 

Green Collar Solutions Limited is expanding into new industries to diversify its portfolio. Among those being considered by the company's stakeholders is a venture into the Aviation industries (purchasing and operating aircfats for commercial and private enterprises).My task in this project is to answer the following questions for the stakeholders.

(i). What are the risk factors associated with different aircrafts?
(ii). Do the accidents rates depend on whether the aircraft is commercial or private?
(iii). Does the model of an aircraft minimize or increase it's risk factors?

Answers to these questions will determine which model,where privately or commercially operated, has the lowest risks of getting into an accident regardless of other factors like weather or human error etc.

# (b) Metrics for success

Being successifully able to analyse the data available and giving well researched information that provides actionable insights to green collar solutions to confidently venture into the Aviation industries while minimizing among others,financial risks.

# (c) Steps for accomplishing the project's objectives.


1. Data collection>>> get data from NTSB aviation accidents report for 1962-2023
2. Data cleaning or preprocessing data using pandas libraries in Python
3. Data exploration - doing an indepth data analysis of the report to show trends,factor relationships and patterns(EDA)
4. Actionable insights and recommendations - translate the findings of the Exploratory data analysis into actionable insights and recommendations.
5. Effective visualizations to communicate the insights and recommendations to Green collar Solutions limited stakeholders to help them in making informed decisions on their new venture.

It's important to note also that i will consider the relevance of the data in the NTSB Report in relation to the task i want to accomplish. Data also has to be complete,accurate,consistent and reliable.

# 2. Import relevants libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline 

# 3. Loading the data into a dataframe

In [2]:
#load  'Aviation_data.csv' file into dataframe 'Aviation_df'

Aviation_df = pd.read_csv('AviationData.csv',encoding='ISO-8859-1',low_memory=False)
#display the first 5 rows of the dataframe
Aviation_df.head()

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country   Latitude   Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0          NaN  ...          Personal         NaN                  2.0   
1          NaN  ...          Personal         NaN                  4.0   
2          NaN  ...          Personal         NaN                  3.0   
3          NaN  ...          Personal         NaN                  2.0   
4          NaN  ...          Personal         NaN                  1.0   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                    0.0                  0.0             0.0   
1                    0.0                  0.0             0.0   
2                    NaN                  NaN             NaN   
3                    0.0                  0.0             0.0   
4                    2.0                  NaN             0.0   

  Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0               UNK                 Cruise  Probable Cause              NaN  
1               UNK                Unknown  Probable Cause       19-09-1996  
2               IMC                 Cruise  Probable Cause       26-02-2007  
3               IMC                 Cruise  Probable Cause       12-09-2000  
4               VMC               Approach  Probable Cause       16-04-1980  

[5 rows x 31 columns]

# Data understanding

1. Data exploration - the structure of the data frame,its features,the datatypes it contains,does it have any missing values,what variables in the dataframe are relevant to our project in risk assessment

In [3]:
Aviation_df.columns

Index(['Event.Id', 'Investigation.Type', 'Accident.Number', 'Event.Date',
       'Location', 'Country', 'Latitude', 'Longitude', 'Airport.Code',
       'Airport.Name', 'Injury.Severity', 'Aircraft.damage',
       'Aircraft.Category', 'Registration.Number', 'Make', 'Model',
       'Amateur.Built', 'Number.of.Engines', 'Engine.Type', 'FAR.Description',
       'Schedule', 'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Report.Status',
       'Publication.Date'],
      dtype='object')

In [4]:
#Shape of the dataframe
Aviation_df.shape

(88889, 31)

In [5]:
#datatypes and missing values
print(Aviation_df.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50249 non-null  object 
 9   Airport.Name            52790 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87572 non-null  object 
 14  Make                    88826 non-null

This dataframe has 88889 rows and 31 columns. a number of columns have some missing values but we will explore this later. It has mixed datatypes (float and object) for numbers and strings respectively. it is of class type:pandas.Core.frame.Dataframe

In [6]:
#Get the descriptive statistics for the dataframe
Aviation_df.describe(include='all')

Event.Id Investigation.Type Accident.Number  Event.Date  \
count            88889              88889           88889       88889   
unique           87951                  2           88863       14782   
top     20001212X19172           Accident      WPR23LA045  1982-05-16   
freq                 3              85015               2          25   
mean               NaN                NaN             NaN         NaN   
std                NaN                NaN             NaN         NaN   
min                NaN                NaN             NaN         NaN   
25%                NaN                NaN             NaN         NaN   
50%                NaN                NaN             NaN         NaN   
75%                NaN                NaN             NaN         NaN   
max                NaN                NaN             NaN         NaN   

             Location        Country Latitude Longitude Airport.Code  \
count           88837          88663    34382     34373        50249   
unique          27758            219    25589     27154        10375   
top     ANCHORAGE, AK  United States  332739N  0112457W         NONE   
freq              434          82248       19        24         1488   
mean              NaN            NaN      NaN       NaN          NaN   
std               NaN            NaN      NaN       NaN          NaN   
min               NaN            NaN      NaN       NaN          NaN   
25%               NaN            NaN      NaN       NaN          NaN   
50%               NaN            NaN      NaN       NaN          NaN   
75%               NaN            NaN      NaN       NaN          NaN   
max               NaN            NaN      NaN       NaN          NaN   

       Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
count         52790  ...             82697       16648         77488.000000   
unique        24871  ...                26       13590                  NaN   
top         Private  ...          Personal       Pilot                  NaN   
freq            240  ...             49448         258                  NaN   
mean            NaN  ...               NaN         NaN             0.647855   
std             NaN  ...               NaN         NaN             5.485960   
min             NaN  ...               NaN         NaN             0.000000   
25%             NaN  ...               NaN         NaN             0.000000   
50%             NaN  ...               NaN         NaN             0.000000   
75%             NaN  ...               NaN         NaN             0.000000   
max             NaN  ...               NaN         NaN           349.000000   

       Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
count            76379.000000         76956.000000    82977.000000   
unique                    NaN                  NaN             NaN   
top                       NaN                  NaN             NaN   
freq                      NaN                  NaN             NaN   
mean                 0.279881             0.357061        5.325440   
std                  1.544084             2.235625       27.913634   
min                  0.000000             0.000000        0.000000   
25%                  0.000000             0.000000        0.000000   
50%                  0.000000             0.000000        1.000000   
75%                  0.000000             0.000000        2.000000   
max                161.000000           380.000000      699.000000   

       Weather.Condition  Broad.phase.of.flight   Report.Status  \
count              84397                  61724           82508   
unique                 4                     12           17075   
top                  VMC                Landing  Probable Cause   
freq               77303                  15428           61754   
mean                 NaN                    NaN             NaN   
std                  NaN                    NaN             NaN   
min          

2. Data quality assessment - check for missing values,duplicates

In [7]:
#display athe total number of missing values 
print(Aviation_df.isnull().sum().sum())

564742


Aviation dataframe has a total of 564,742 missing values

In [8]:
#let's see the total number of missing values from each column
print(Aviation_df.isnull().sum())


Event.Id                      0
Investigation.Type            0
Accident.Number               0
Event.Date                    0
Location                     52
Country                     226
Latitude                  54507
Longitude                 54516
Airport.Code              38640
Airport.Name              36099
Injury.Severity            1000
Aircraft.damage            3194
Aircraft.Category         56602
Registration.Number        1317
Make                         63
Model                        92
Amateur.Built               102
Number.of.Engines          6084
Engine.Type                7077
FAR.Description           56866
Schedule                  76307
Purpose.of.flight          6192
Air.carrier               72241
Total.Fatal.Injuries      11401
Total.Serious.Injuries    12510
Total.Minor.Injuries      11933
Total.Uninjured            5912
Weather.Condition          4492
Broad.phase.of.flight     27165
Report.Status              6381
Publication.Date          13771
dtype: i

The only columns with no missing values are; 
Event.Id                      
Investigation.Type            
Accident.Number               
Event.Date

In [9]:
#checking for duplicated rows within the dataframe
duplicate_rows = Aviation_df[Aviation_df.duplicated()]
duplicate_rows

Empty DataFrame
Columns: [Event.Id, Investigation.Type, Accident.Number, Event.Date, Location, Country, Latitude, Longitude, Airport.Code, Airport.Name, Injury.Severity, Aircraft.damage, Aircraft.Category, Registration.Number, Make, Model, Amateur.Built, Number.of.Engines, Engine.Type, FAR.Description, Schedule, Purpose.of.flight, Air.carrier, Total.Fatal.Injuries, Total.Serious.Injuries, Total.Minor.Injuries, Total.Uninjured, Weather.Condition, Broad.phase.of.flight, Report.Status, Publication.Date]
Index: []

[0 rows x 31 columns]

Aviation dataframe has no duplicated rows

# Data Cleaning

# dealing with missing values


In [10]:
#first let's look at the dataframe again
Aviation_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50249 non-null  object 
 9   Airport.Name            52790 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87572 non-null  object 
 14  Make                    88826 non-null

In [11]:
# now we drop columns that have alot of missing values and have little or no relevance to our task
# first make a copy of the original dataframe for later reference if needed
Aviation_df_copy = Aviation_df.copy()
Aviation_df_copy

Event.Id Investigation.Type Accident.Number  Event.Date  \
0      20001218X45444           Accident      SEA87LA080  1948-10-24   
1      20001218X45447           Accident      LAX94LA336  1962-07-19   
2      20061025X01555           Accident      NYC07LA005  1974-08-30   
3      20001218X45448           Accident      LAX96LA321  1977-06-19   
4      20041105X01764           Accident      CHI79FA064  1979-08-02   
...               ...                ...             ...         ...   
88884  20221227106491           Accident      ERA23LA093  2022-12-26   
88885  20221227106494           Accident      ERA23LA095  2022-12-26   
88886  20221227106497           Accident      WPR23LA075  2022-12-26   
88887  20221227106498           Accident      WPR23LA076  2022-12-26   
88888  20221230106513           Accident      ERA23LA097  2022-12-29   

              Location        Country   Latitude   Longitude Airport.Code  \
0      MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1       BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2        Saltville, VA  United States  36.922223  -81.878056          NaN   
3           EUREKA, CA  United States        NaN         NaN          NaN   
4           Canton, OH  United States        NaN         NaN          NaN   
...                ...            ...        ...         ...          ...   
88884    Annapolis, MD  United States        NaN         NaN          NaN   
88885      Hampton, NH  United States        NaN         NaN          NaN   
88886       Payson, AZ  United States    341525N    1112021W          PAN   
88887       Morgan, UT  United States        NaN         NaN          NaN   
88888       Athens, GA  United States        NaN         NaN          NaN   

      Airport.Name  ... Purpose.of.flight         Air.carrier  \
0              NaN  ...          Personal                 NaN   
1              NaN  ...          Personal                 NaN   
2              NaN  ...          Personal                 NaN   
3              NaN  ...          Personal                 NaN   
4              NaN  ...          Personal                 NaN   
...            ...  ...               ...                 ...   
88884          NaN  ...          Personal                 NaN   
88885          NaN  ...               NaN                 NaN   
88886       PAYSON  ...          Personal                 NaN   
88887          NaN  ...          Personal  MC CESSNA 210N LLC   
88888          NaN  ...          Personal                 NaN   

      Total.Fatal.Injuries Total.Serious.Injuries Total.Minor.Injuries  \
0                      2.0                    0.0                  0.0   
1                      4.0                    0.0                  0.0   
2                      3.0                    NaN                  NaN   
3                      2.0                    0.0                  0.0   
4                      1.0                    2.0                  NaN   
...                    ...                    ...                  ...   
88884                  0.0                    1.0                  0.0   
88885                  0.0                    0.0                  0.0   
88886                  0.0                    0.0                  0.0   
88887                  0.0                    0.0                  0.0   
88888                  0.0                    1.0                  0.0   

      Total.Uninjured Weather.Condition  Broad.phase.of.flight  \
0                 0.0               UNK                 Cruise   
1                 0.0               UNK                Unknown   
2                 NaN               IMC                 Cruise   
3                 0.0               IMC                 Cruise   
4                 0.0               VMC               Approach   
...               ...               ...                    ...   
88884             0.0               NaN                    NaN   
88885             0.0    

In [12]:
Aviation_df = Aviation_df_copy
Aviation_df

Event.Id Investigation.Type Accident.Number  Event.Date  \
0      20001218X45444           Accident      SEA87LA080  1948-10-24   
1      20001218X45447           Accident      LAX94LA336  1962-07-19   
2      20061025X01555           Accident      NYC07LA005  1974-08-30   
3      20001218X45448           Accident      LAX96LA321  1977-06-19   
4      20041105X01764           Accident      CHI79FA064  1979-08-02   
...               ...                ...             ...         ...   
88884  20221227106491           Accident      ERA23LA093  2022-12-26   
88885  20221227106494           Accident      ERA23LA095  2022-12-26   
88886  20221227106497           Accident      WPR23LA075  2022-12-26   
88887  20221227106498           Accident      WPR23LA076  2022-12-26   
88888  20221230106513           Accident      ERA23LA097  2022-12-29   

              Location        Country   Latitude   Longitude Airport.Code  \
0      MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1       BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2        Saltville, VA  United States  36.922223  -81.878056          NaN   
3           EUREKA, CA  United States        NaN         NaN          NaN   
4           Canton, OH  United States        NaN         NaN          NaN   
...                ...            ...        ...         ...          ...   
88884    Annapolis, MD  United States        NaN         NaN          NaN   
88885      Hampton, NH  United States        NaN         NaN          NaN   
88886       Payson, AZ  United States    341525N    1112021W          PAN   
88887       Morgan, UT  United States        NaN         NaN          NaN   
88888       Athens, GA  United States        NaN         NaN          NaN   

      Airport.Name  ... Purpose.of.flight         Air.carrier  \
0              NaN  ...          Personal                 NaN   
1              NaN  ...          Personal                 NaN   
2              NaN  ...          Personal                 NaN   
3              NaN  ...          Personal                 NaN   
4              NaN  ...          Personal                 NaN   
...            ...  ...               ...                 ...   
88884          NaN  ...          Personal                 NaN   
88885          NaN  ...               NaN                 NaN   
88886       PAYSON  ...          Personal                 NaN   
88887          NaN  ...          Personal  MC CESSNA 210N LLC   
88888          NaN  ...          Personal                 NaN   

      Total.Fatal.Injuries Total.Serious.Injuries Total.Minor.Injuries  \
0                      2.0                    0.0                  0.0   
1                      4.0                    0.0                  0.0   
2                      3.0                    NaN                  NaN   
3                      2.0                    0.0                  0.0   
4                      1.0                    2.0                  NaN   
...                    ...                    ...                  ...   
88884                  0.0                    1.0                  0.0   
88885                  0.0                    0.0                  0.0   
88886                  0.0                    0.0                  0.0   
88887                  0.0                    0.0                  0.0   
88888                  0.0                    1.0                  0.0   

      Total.Uninjured Weather.Condition  Broad.phase.of.flight  \
0                 0.0               UNK                 Cruise   
1                 0.0               UNK                Unknown   
2                 NaN               IMC                 Cruise   
3                 0.0               IMC                 Cruise   
4                 0.0               VMC               Approach   
...               ...               ...                    ...   
88884             0.0               NaN                    NaN   
88885             0.0    

In [15]:
# start by dropping columns with most missing values and does not cause harm to the data needed for the project
Aviation_df.drop(columns= ['Publication.Date', 'Longitude','Latitude','Registration.Number','FAR.Description'], inplace=True) 

In [17]:
Aviation_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 26 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Airport.Code            50249 non-null  object 
 7   Airport.Name            52790 non-null  object 
 8   Injury.Severity         87889 non-null  object 
 9   Aircraft.damage         85695 non-null  object 
 10  Aircraft.Category       32287 non-null  object 
 11  Make                    88826 non-null  object 
 12  Model                   88797 non-null  object 
 13  Amateur.Built           88787 non-null  object 
 14  Number.of.Engines       82805 non-null

# imputing

In [28]:
#replace the missing number of values in the 'number of engines' column with median
columns_to_fill = ['Number.of.Engines','Total.Fatal.Injuries','Total.Serious.Injuries','Total.Minor.Injuries','Total.Uninjured', inplace=True]
for col in columns_to_fill:
    Aviation_df[col] = Aviation_df[col].fillna(Aviation_df[col].median())

In [33]:
cols_to_fill = ['Location','Country','Airport.Code','Airport.Name','Injury.Severity','Aircraft.damage','Aircraft.Category','Make','Model','Amateur.Built','Engine.Type','Schedule','Purpose.of.flight','Air.carrier','Weather.Condition','Broad.phase.of.flight','Report.Status']
for col in cols_to_fill:
    if col in Aviation_df.columns:
        Aviation_df[col].fillna('unknown', inplace=True)
    

In [34]:
#assert that there are no missing values in the dataframe
assert Aviation_df.isnull().sum().sum()== 0

In [37]:
#ensure dates are in the correct format
Aviation_df['Event.Date'] = pd.to_datetime(Aviation_df['Event.Date'])